cyral/resource_cyral_listener.go

```diff
@@ -0,0 +1,276 @@
+package cyral
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/cyralinc/terraform-provider-cyral/client"
+)
+
+// SidecarListener struct for sidecar listener.
+type SidecarListener struct {
+	SidecarId        string            `json:"-"`
+	ListenerId       string            `json:"id"`
+	RepoTypes        []string          `json:"repoTypes"`
+	UnixListener     *UnixListener     `json:"unixListener,omitempty"`
+	TcpListener      *TcpListener      `json:"tcpListener,omitempty"`
+	Multiplexed      bool              `json:"multiplexed,omitempty"`
+	MysqlSettings    *MysqlSettings    `json:"mysqlSettings,omitempty"`
+	S3Settings       *S3Settings       `json:"s3Settings,omitempty"`
+	DynamoDbSettings *DynamoDbSettings `json:"dynamoDBSettings,omitempty"`
+}
+type UnixListener struct {
+	File string `json:"file"`
+}
+type TcpListener struct {
+	Host string `json:"host,omitempty"`
+	Port int    `json:"port"`
+}
+type MysqlSettings struct {
+	DbVersion    string `json:"dbVersion,omitempty"`
+	CharacterSet string `json:"characterSet,omitempty"`
+}
+type S3Settings struct {
+	ProxyMode bool `json:"proxyMode,omitempty"`
+}
+type DynamoDbSettings struct {
+	ProxyMode bool `json:"proxyMode,omitempty"`
+}
+
+var ReadSidecarListenersConfig = ResourceOperationConfig{
+	Name:       "SidecarListenersResourceRead",
+	HttpMethod: http.MethodGet,
+	CreateURL: func(d *schema.ResourceData, c *client.Client) string {
+		return fmt.Sprintf("https://%s/v1/sidecars/%s/listeners/%s",
+			c.ControlPlane,
+			d.Get("sidecar_id").(string),
+			d.Id())
+	},
+	NewResponseData: func(_ *schema.ResourceData) ResponseData { return &ReadSidecarListenersAPIResponse{} },
+}
+
+type ReadSidecarListenersAPIResponse struct {
+	Listener *SidecarListener `json:"listener"`
+}
+type CreateListenerAPIResponse struct {
+	ListenerId string `json:"listenerId"`
+}
+
+func (c CreateListenerAPIResponse) WriteToSchema(d *schema.ResourceData) error {
+	d.SetId(c.ListenerId)
```

Comment:
```suggestion
	d.SetId(marshalComposedID([]string{d.Get("sidecar_id"), c.ListenerId}, "/"))
```

It is best to keep a consistency between the ID the user uses to import the resource in:

```
terraform import {sidecar_id}/{listener_id}
```

and the actual ID of the Terraform resource. I believe the average Terraform user would expect this consistency.

`terraform import -h` displays the following message:

```
Usage: terraform [global options] import [options] ADDR ID

...

  The ADDR specified is the address to import the resource to. Please
  see the documentation online for resource addresses. The ID is a
  resource-specific ID to identify that resource being imported. Please
  reference the documentation for the resource type you're importing to
  determine the ID syntax to use. It typically matches directly to the ID
  that the provider uses.
```